### PR TITLE
feat(NODE-3697): reduce serverSession allocation

### PR DESCRIPTION
**Pull request by nbbeeken**
#### What is changing?

- `incrementTransactionNumber` will no longer acquire a serverSession because it won't use the getter

#### What is the motivation for this change?

With this change our driver will only acquire a serverSession when the getter is used which is in applySession, used inside the Connection class. This limits the number of serverSession allocated to be bounded by at least the number of concurrent operations running and at best by the max number of connection in our connectionPool.

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the correct format: `<type>(NODE-xxxx)<!>: <description>`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

### CODE REVIEW

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: test/integration/sessions/sessions.spec.prose.test.ts
Comment (nbbeeken) [reply test/integration/sessions/sessions.spec.prose.test.ts:32]: (talked offline) Might as well still catch the case where these fail, succeeding is parallel to the test's goal but is a nice confirmation. Also I think we may still want to inspect the error cases, what if the error relates to improper session usage issues.

File: src/sessions.ts
Comment (nbbeeken) [reply src/sessions.ts:987]: Deleted!

File: src/sessions.ts
Comment (nbbeeken) [reply src/sessions.ts:270]: I've added hasEnded logic to the getter but this code is still necessary to attach the kServerSession for inspecting post usage.

File: src/sessions.ts
Comment (nbbeeken) [reply src/sessions.ts:969]: Willing to do this, lets hold off til team review

**Comment by dariakp** [comment]
I'm sorry I haven't gotten around to a review yet, but it's on my list to do Monday!

**Review by dariakp — CHANGES_REQUESTED** [review]

File: src/sessions.ts
Comment (dariakp) [reply src/sessions.ts:369]: Just re-read this again, and I see we want to store the txnNumber on the session before acquiring the server session because we don't want to acquire the server session here. Can we update the description comment above explaining this?

File: src/sessions.ts
Comment (nbbeeken) [reply src/sessions.ts:45]: It was never possible for `session.serverSession` to be null the getter calls acquire if the the symbol is nullish and acquire always returns a ServerSession.

File: test/integration/sessions/sessions.spec.prose.test.ts
Comment (nbbeeken) [reply test/integration/sessions/sessions.spec.prose.test.ts:22]: Done and I've updated the wording

File: src/sessions.ts
Comment (nbbeeken) [reply src/sessions.ts:369]: Updated!

File: test/unit/sessions.test.js
Comment (nbbeeken) [reply test/unit/sessions.test.js]: Added a test for the opposite condition

File: src/sessions.ts
Comment (dariakp) [reply src/sessions.ts:45]: Alright, I'll buy it lol

File: src/sessions.ts
Comment (dariakp) [reply src/sessions.ts:166]: I don't really agree on the `!!`, because `!!` is entirely equivalent to `Boolean` and unlike the other examples on the list, there aren't multiple interpretations to what a `!` in front of a variable could mean, it's just shorter. However, if the entire codebase used `Boolean` and you decided to do `!!`, I would give the reverse feedback, because in matters of opinion, the best practice is to opt for *consistency*. I am not opposed to explicit casts in principle, if that's what the entire team prefers, but if we want to make this change in convention, then the change should be applied to the entire codebase and enforced with the lint rule you are referencing.

File: src/sessions.ts
Comment (dariakp) [reply src/sessions.ts:193]: If we are confident that `this[kServerSession]` cannot be undefined here, then we can safely add a `if !this[kServerSession] { throw new MongoRuntimeError('...') }`. We shouldn't be using ts expect error in production code.

As for nullish values in general, you didn't really answer my question about why we can't initialize `this[kServerSession]` to null instead of undefined  and avoid any need for `ServerSession | undefined` typings altogether.

File: test/unit/sessions.test.js
Comment (dariakp) [inline test/unit/sessions.test.js:234]: should it only acquire it if it is not set? I feel like there's 2 cases here, one if it's not set, and one if it's already defined, i.e., if we set `session[serverSessionSymbol] = 'some random value'`, then we can expect `session.serverSession` to equal that random value

File: test/unit/sessions.test.js
Comment (dariakp) [inline test/unit/sessions.test.js:254]: nit: insert blank line above

**Review by dariakp — CHANGES_REQUESTED** [review]
Much easier to follow the tests now, thanks! I resolved the comments that have been addressed, there are 2 comments with active discussions left, and I noticed a couple more nits here:

File: src/sessions.ts
Comment (nbbeeken) [reply src/sessions.ts:78]: The benefit of the doc here is you get to see it when you hover on the property in VSCode, whch you normally get for string props but not for symbol props

File: src/sessions.ts
Comment (nbbeeken) [reply src/sessions.ts]: Added throw

File: test/unit/sessions.test.js
Comment (nbbeeken) [reply test/unit/sessions.test.js:234]: tests changed cus of throw

File: src/sessions.ts
Comment (nbbeeken) [reply src/sessions.ts:972]: @baileympearson let's file a follow up here, instead of using this PR

Comment (baileympearson) [reply src/sessions.ts:968]: @nbbeeken Sounds good, we can focus on getting this PR merged

File: test/unit/sessions.test.js
Comment (dariakp) [inline test/unit/sessions.test.js:225]: Now that we a clearly defined behavior for the getter, I think we should split up the cases to cover the following scenarios:
- session symbol is non-null, the session is explicit, regardless of session ended or not
- session symbol is non-null, the session is implicit, regardless of session ended or not
- session symbol is null, the session is explicit, regardless of session ended or not
- session symbol is null, the session is implicit, session not ended
- session symbol is null, the session is implicit, session is ended

**Review by dariakp — CHANGES_REQUESTED** [review]

File: test/unit/sessions.test.js
Comment (baileympearson) [inline test/unit/sessions.test.js:180]:
```suggestion
const topology = new Topology('localhost:27017', {});
```

File: test/unit/sessions.test.js
Comment (baileympearson) [inline test/unit/sessions.test.js:187]:
```suggestion
const topology = new Topology('localhost:27017', {});
```

File: test/unit/sessions.test.js
Comment (dariakp) [inline test/unit/sessions.test.js:259]:
```suggestion
        it('should throw if the session ended before serverSession was acquired', () => {
```

File: test/unit/sessions.test.js
Comment (dariakp) [inline test/unit/sessions.test.js:265]:
```suggestion
        it('should acquire a serverSession if clientSession.hasEnded is false and serverSession is not set', () => {
```

File: test/unit/sessions.test.js
Comment (dariakp) [inline test/unit/sessions.test.js:273]: We're missing a test case where `it('should return the existing serverSession and not acquire a new one if one is already set')`

Comment (nbbeeken) [reply test/unit/sessions.test.js:290]: Added!

File: test/unit/sessions.test.js
Comment (dariakp) [inline test/unit/sessions.test.js:274]: can you make this test the case where the session has ended, too?

**Review by dariakp — CHANGES_REQUESTED** [review]

File: test/unit/sessions.test.js
Comment (nbbeeken) [reply test/unit/sessions.test.js:274]: done!

**Review by dariakp — APPROVED** [review]
LGTM assuming CI passes!

**Pull request merged**
